apps/web/components/apps/AppPage.tsx

```diff
@@ -70,22 +73,63 @@ export const AppPage = ({
   dependencies,
   concurrentMeetings,
   paid,
-  dirName,
 }: AppPageProps) => {
   const { t, i18n } = useLocale();
+  const router = useRouter();
   const searchParams = useCompatSearchParams();
 
   const hasDescriptionItems = descriptionItems && descriptionItems.length > 0;
 
-  const mutation = useAddAppMutation(null, {
-    onSuccess: (data) => {
-      if (data?.setupPending) return;
-      showToast(t("app_successfully_installed"), "success");
-    },
-    onError: (error) => {
-      if (error instanceof Error) showToast(error.message || t("app_could_not_be_installed"), "error");
-    },
-  });
+  const mutation = useAddAppMutation(null);
+
+  const [isLoading, setIsLoading] = useState<boolean>(mutation.isPending);
+
+  const handleAppInstall = () => {
+    setIsLoading(true);
+    if (isConferencing(categories)) {
+      mutation.mutate(
+        {
+          isOmniInstall: true,
+          type,
+          variant,
+          slug,
+          returnTo:
+            WEBAPP_URL +
+            getAppOnboardingUrl({
+              slug,
+              step: AppOnboardingSteps.EVENT_TYPES_STEP,
+            }),
+        },
+        {
+          onSuccess: (data) => {
+            if (data?.setupPending) return;
+            // for non-oAuth apps
+            router.push(
+              getAppOnboardingUrl({
+                slug,
+                step: AppOnboardingSteps.EVENT_TYPES_STEP,
+              })
+            );
+            showToast(t("app_successfully_installed"), "success");
+          },
+          onError: (error) => {
+            if (error instanceof Error) showToast(error.message || t("app_could_not_be_installed"), "error");
+            setIsLoading(false);
+          },
+        }
+      );
+    } else if (
+      !doesAppSupportTeamInstall({
+        appCategories: categories,
+        concurrentMeetings: concurrentMeetings,
+        isPaid: !!paid,
+      })
+    ) {
+      mutation.mutate({ type });
+    } else {
+      router.push(getAppOnboardingUrl({ slug, step: AppOnboardingSteps.ACCOUNTS_STEP }));
+    }
+  };
```

Comment:
Two cents here. Why not let the server handle all this logic? I'm noticing that we redirect to different places depending of certain conditions. Why not send all the data to the server and let the server return a `returnTo` URL according to the conditions that are run on the backend? IMO having this much logic on the frontend seems like a code smell. Thoughts on this @emrysal @keithwillcode ?

Reply:
can we do it in a seprate PR, as a lot of things could go wrong if we decide to refactor `useAddAppMutation` function
https://github.com/calcom/cal.com/pull/15193#discussion_r1624835882

Reply:
tried a fix here, let me know what you think

https://github.com/calcom/cal.com/pull/15193/commits/b2f478cdb68ab0c5caf30c581e10547548bafbbd and https://github.com/calcom/cal.com/pull/15193/commits/69378039ec5df39cd7e74d39bf4c3fd148ac537b

@joeauyeung @zomars